src/system/random/basic.lean

```diff
@@ -0,0 +1,467 @@
+do g ← io.mk_generator,
+   return $ (cmd.run ⟨g⟩).1
+
+section random
+variables [preorder α] [random α]
+
+/-- randomly generate a value of type α -/
+def random : io α :=
+io.run_rand (rand.random α)
+
+/-- randomly generate an infinite series of value of type α -/
+def random_series : io (stream α) :=
+io.run_rand (rand.random_series α)
+
+end random
+
+section bounded_random
+variables [preorder α] [bounded_random α]
+
+/-- randomly generate a value of type α between `x` and `y` -/
+def random_r (x y : α) (p : x ≤ y) : io (x .. y) :=
+io.run_rand (bounded_random.random_r _ x y p)
+
+/-- randomly generate an infinite series of value of type α between `x` and `y` -/
+def random_series_r (x y : α) (h : x ≤ y) : io (stream $ x .. y) :=
+io.run_rand (rand.random_series_r x y h)
+
+end bounded_random
+
+end io
+
+namespace tactic
+
+/-- create a seeded random number generator in the `tactic` monad -/
+meta def mk_generator : tactic std_gen := do
+tactic.unsafe_run_io @io.mk_generator
+
+/-- run `cmd` using the a randomly seeded random number generator
+in the tactic monad -/
+meta def run_rand {α : Type u} (cmd : rand α) : tactic α := do
+⟨g⟩ ← tactic.up mk_generator,
+return (cmd.run ⟨g⟩).1
+
+variables {α : Type u}
+
+section bounded_random
+variables [preorder α] [bounded_random α]
+
+/-- use `random_r` in the `tactic` monad -/
+meta def random_r (x y : α) (h : x ≤ y) : tactic (x .. y) :=
+run_rand (rand.random_r x y h)
+
+/-- use `random_series_r` in the `tactic` monad -/
+meta def random_series_r (x y : α) (h : x ≤ y) : tactic (stream $ x .. y) :=
+run_rand (rand.random_series_r x y h)
+
+end bounded_random
+
+section random
+
+variables [preorder α] [random α]
+
+/-- use `random` in the `tactic` monad -/
+meta def random : tactic α :=
+run_rand (rand.random α)
+
+/-- use `random_series` in the `tactic` monad -/
+meta def random_series : tactic (stream α) :=
+run_rand (rand.random_series α)
+
+end random
+
+end tactic
+
+namespace bool
+
+/-- Make `i` into an element of the interval `x .. y` if feasible
+and return an arbitrary element of `x .. y` otherwise -/
+def fit_in_Icc (x y : bool) (p : x ≤ y) (i : bool) : x .. y := do
+  if hx : x ≤ i ∧ i ≤ y
+  then ⟨ i, hx ⟩
+  else ⟨ x , le_refl x , p ⟩
+
+open ulift (hiding inhabited)
+variables {g : Type} [random_gen g]
+
+/-- generate a randomly generated boolean value -/
+protected def get_random : rand_g g bool :=
+⟨ prod.map id up ∘ @rand_bool g _ ∘ down ⟩
+
+end bool
+
+instance : random bool :=
+{ random   := λ g, @bool.get_random _,
+  random_r := λ g _inst x y p, bool.fit_in_Icc _ _ p <$> (@bool.get_random g _inst) }
+
+open nat (succ one_add mod_eq_of_lt zero_lt_succ add_one succ_le_succ)
+
+namespace stream
+
+variable {α : Type u}
+
+open list (length) stream (approx)
+
+lemma length_approx :
+  ∀ (s : stream α) (n : ℕ), length (approx n s) = n
+| s 0 := rfl
+| s (succ n) := by simp [approx,length,one_add,length_approx]
+
+end stream
+
+variables {g : Type} [random_gen g]
+
```

Comment:
This is probably the slowest way to produce random bit vectors that I know of.  We should create a random natural number, and then convert it to a bit vector.

Reply:
improved